Add OpenTracing compatibility section. (#1101)
open-telemetry/opentelemetry-specification: specification/compatibility/opentracing.md

# OpenTracing Compatibility

<details>
<summary>Table of Contents</summary>

* [Abstract](#abstract)
* [Create an OpenTracing Tracer Shim](#create-an-opentracing-tracer-shim)
* [Tracer Shim](#tracer-shim)
  * [Inject](#inject)
  * [Extract](#extract)
* [Span Shim](#span-shim)
  * [OpenTelemetry Span and SpanContext relationship](#opentelemetry-span-and-spancontext-relationship)
  * [Get Context](#get-context)
  * [Get Baggage Item](#get-baggage-item)
  * [Set Baggage Item](#set-baggage-item)
  * [Set Tag](#set-tag)
  * [Log](#log)
  * [Finish](#finish)
  * [SpanContext Shim](#spancontext-shim)
    * [Get Baggage Items](#get-baggage-items)
* [ScopeManager Shim](#scopemanager-shim)
  * [Activate a Span](#activate-a-span)
  * [Get the active Span](#get-the-active-span)
* [Semantic conventions mapping](#semantic-conventions-mapping)

</details>

## Abstract

The OpenTelemetry project aims to provide backwards compatibility with the
[OpenTracing](https://opentracing.io) project in order to ease migration of

> instrumented codebases.

Author: Actually I mention it below: https://github.com/open-telemetry/opentelemetry-specification/pull/1101/files#diff-efb9735083a18d449551326756460206cf789d785e16e4a7301c9eb3e0205f36R46

of any SDK.

More specifically, the intention is to allow OpenTracing instrumentation to be
recorded using OpenTelemetry. This Shim Layer MUST NOT publicly expose any
upstream OpenTelemetry API.

This functionality MUST be defined in its own OpenTracing Shim Layer, not in the
OpenTracing nor the OpenTelemetry API or SDK.

The OpenTracing Shim and the OpenTelemetry API/SDK are expected to be consumed
simultaneously in a running service, in order to ease migration from the former
to the latter.

## Create an OpenTracing Tracer Shim

This operation is used to create a new OpenTracing `Tracer`:

This operation MUST accept the following parameters:

- An OpenTelemetry `Tracer`, used to create `Span`s.
- A set of OpenTelemetry `Propagator`s of the supported types, used to perform
  context injection and extraction. Usually these are the global
  `Composite Propagator`s.

The API MUST return an OpenTracing `Tracer`.

## Tracer Shim

### Inject

Parameters:

- A `SpanContext`.
- A `Format` descriptor.
- A carrier.

Inject the underlying OpenTelemetry `Span` and `Bagagge` using the registered
OpenTelemetry `Propagator`s:

> - `TEXT_MAP` and `HTTP_HEADERS` formats MUST use the registered OpenTelemetry

Author: Is [this](https://github.com/jaegertracing/jaeger-client-java/blob/master/jaeger-core/src/main/java/io/jaegertracing/internal/propagation/TextMapCodec.java#L203) the encoding you mention? If yes, I created a prototype that allows specifying encoding/decoding: https://github.com/open-telemetry/opentelemetry-java/compare/master...carlosalberto:ot_shim_encoded?expand=1

But if we ever support this encoding in our OTel Jaeger Propagators, then we would be repeating the conversion, which sounds bad. In that case, we could go with the approach you mentioned in another comment:

```go
otTracer := createTracerShim(otelTracer, propagators = {
    ot.TEXT_MAP: otelsdk.getJaegerTextMapPropagator(),
    ot.HTTP_HEADERS: otel.getHTTPTextPropagator(),
    ot.BINARY: null,
})
```

If not specified, we simply use OTel TextMap for both TextMap/HttpHeaders. Any opinion on this one? I'm fine with either of them.

> - `TEXT_MAP` and `HTTP_HEADERS` formats MUST use the registered OpenTelemetry

Reviewer: Not quite following why we'd be repeating conversion - OT shim won't do anything by itself, just delegate to OTEL propagator. I am just saying that there needs to be a way to use different propagators for TEXT_MAP vs. HTTP_HEADERS formats, contrary to what this L76 says.

Your branch is kind of going in this direction, but not quite, because this:

```java
public static io.opentracing.Tracer createTracerShim(
      OpenTelemetry openTelemetry, boolean urlEncoding) {
```

_fixes_ the urlEncoding for a given OT tracer. But the whole point is that whether to apply urlEncoding or not is not a fixed property of the tracer, but the side effect of which format the user is trying use to to encode the context. A single tracer can be used in a service that has both HTTP server _and_ publishes to Kafka, and that service would use different OT formats for those two trace points.

Passing a map of propagators instead of a boolean flag when creating a shim should resolve this generically.

> - `TEXT_MAP` and `HTTP_HEADERS` formats MUST use the registered OpenTelemetry

Author: Hey @yurishkuro I'm fine allowing the user to pass a set of propagators at OT Shim creation time - then the behavior should be IMHO:

a) If user specified any Propagator, have the Shim use them for **all** propagation, ignoring the global Propagators.
b) If the user didn't specify any Propagator, simply fallback to use the global Propagators.

Is that a good compromise for you? Or think that for a) we should _also_ override the global Propagators?

> - `TEXT_MAP` and `HTTP_HEADERS` formats MUST use the registered OpenTelemetry

Reviewer: I don't think creating a shim should affect global propagators.

I would combine a) and b) into this:
  * user can pass specific propagators for each of the OpenTracing formats
  * if a propagator is not specified for one of the OT formats, the shim will fall back to default OTEL propagator as follows:
    * OT.TEXT_MAP -> OTEL.HTTP  
    * OT.HTTP_HEADERS -> OTEL.HTTP
    * OT.BINARY -> not supported at this time

There is also a question on the typing: in OTEL the text and binary propagators are likely to have different signatures.

> - `TEXT_MAP` and `HTTP_HEADERS` formats MUST use the registered OpenTelemetry

Author: Thanks, will update the appropriately, along with a small prototype for this (just as done before).

  `HTTPTextPropagator`, if any.

> Errors MUST NOT be raised if the specified `Format` is `BINARY`.

Author: >  if a propagator is not configured for a certain OT format, then a fallback behavior can be prescribed.

Makes sense. 

>  e.g. in Go an error is explicitly returned:

Good call. Will adjust to allow this.

### Extract

Parameters:

- A `Format` descriptor.
- A carrier.

Extract OpenTelemetry `Span` and `Baggage` from a carrier using the registered
OpenTelemetry `Propagator`s:

- `TEXT_MAP` and `HTTP_HEADERS` formats MUST use the registered OpenTelemetry
  `HTTPTextPropagator`, if any.

Returns a `SpanContext` Shim with the underlying extracted OpenTelemetry
`Span` and `Baggage`, or null if either the `Format` is `BINARY` or
no value could be extracted.

## Span Shim

The `Span` operations MUST be implemented using underlying OpenTelemetry `Span`

> and `Baggage` values through a `SpanContext` Shim.

Author: I recurrently mention a **SpanContext Shim** as this is the 'core' object for the `SpanContext`/`Baggage` composite object. I can leave this out but I think there's value in having, at the document level, as a 'first class' citizen, in order to clarify the Shim expectations.

> In order to satisfy the OpenTracing `Span` requirements:

Author: All right, this can be moved to its own section.

  and a `Baggage`. The `SpanContext` Shim MUST be immutable and MUST be
  replaced every time baggage is updated through [Set Baggage Item](#set-baggage-item).
- An underlying OpenTelemetry `Span` MUST be associated with only one
  `SpanContext` Shim at a time. This is done in order to keep its linked `Baggage`
  consistent across all execution units at all times.

This is a simple graphical representation of the mentioned objects:

```
  Span Shim
  +- OpenTelemetry Span
  +- SpanContext Shim
        +- OpenTelemetry SpanContext
        +- OpenTelemetry Baggage
```

The OpenTelemetry `Span` in `Span` Shim is also used to get and set
its currently associated `SpanContext` Shim. See
[OpenTelemetry Span and SpanContext relationship](#opentelemetry-span-and-spancontext-relationship).

The `Log` operations MUST be implemented using the OpenTelemetry
`Span`'s `Add Events` operations.

The `Set Tag` operations MUST be implemented using the OpenTelemetry
`Span`'s `Set Attributes` operations.

### OpenTelemetry Span and SpanContext relationship

A given OpenTelemetry `Span` MUST be associated with only one

> `SpanContext` Shim at all times for all execution units, in order

Author: I think we use it in other places - essentially we wanted to get far from using `threads` as there are similar components such as coroutines in Python.

> `SpanContext` Shim at all times for all execution units, in order

Author: Filled #1511

to keep any linked `Baggage` consistent across all execution
units at all times.

An example of this scenario is having an OpenTracing `Span` have its
[Set Baggage Item](#set-baggage-item) operation called from two different
execution units (e.g. threads), and afterwards have its [Context](#get-context)
fetched in order to [iterate over its baggage values](#get-baggage-items).

Getting and setting the currently associated `SpanContext` for a specified
OpenTelemetry `Span` MUST be safe to call from different execution units.

Managing this relationship is an implementation detail. It can be implemented,
for example, with the help of a global synchronized dictionary, or with an

> additional attribute in OpenTelemetry `Span` objects for dynamic languages.

Author: In Java and Python `Baggage` exists in its own object, which basically means we can leave `Context` entirely out (at least for those languages). 

> Specifically, in OT the span.setBaggage operation is change-in-place and affects what baggage can be seen by different threads

This is indeed the reason I'm mentioning here the details. However, I'm *also* trying to keep the invariant of an actual `SpanContext` being entirely replaced:

```java
// OT perspective
SpanContext ctx1 = span.getContext();
span.setBaggageItem("hey", "you");
assertNotEquals(ctx1, span.getContext());
```

If you think this invariant is not worth keeping (and it's enough to simply keep the *actual values* consistent), I can try to play around and simplify things.

### Get Context

Returns the [associated](#opentelemetry-span-and-spancontext-relationship)
`SpanContext` Shim.

### Get Baggage Item

Parameters:

- The baggage key, a string.

Returns a value for the specified key in the OpenTelemetry `Baggage` of the
associated `SpanContext` Shim or null if none exists.

This is accomplished by getting the
[associated](#opentelemetry-span-and-spancontext-relationship)
`SpanContext` Shim and do a lookup for the specified key in the OpenTelemetry
`Baggage` instance.

### Set Baggage Item

Parameters:

- The baggage key, a string.
- The baggage value, a string.

Creates a new `SpanContext` Shim with a new OpenTelemetry `Baggage` containing
the specified `Baggage` key/value pair. The resulting `SpanContext` Shim is then
[associated](#opentelemetry-span-and-spancontext-relationship) to the underlying
OpenTelemetry `Span`.

### Set Tag

Parameters:

- The tag key, a string.
- The tag value, which must be either a string, a boolean value, or a numeric type.

Calls `Set Attribute` on the underlying OpenTelemetry `Span` with the specified
key/value pair.

The value MUST be [mapped](#semantic-conventions-mapping) to the respective

> OpenTelemetry `Attribute`, or converted to a string if the value type is not supported.

Reviewer: I recommend moving that mapping section right here, because it does not apply to anything else and forces the reader to switch context.

Q: are we going to do anything when `sampling.priority` tag is set?

Author: Not that I'm aware of - we can add it here if you have a clear idea of what should be we mapping it to.

> If the type of the specified value is not supported, the value MUST be converted

Reviewer: > type of the specified value is not supported

Not supported by whom? Would like this to be clearer. In dynamic languages the OTEL SDK would need to deal with any type of value, so the shim does not need to do any additional conversion. It's only when OTEL API in strongly typed languages restricts the types of values for attributes then we may need this to-string coercion (but then the equivalent OT API may not allow `any` value either, e.g. in Java).

Author: Not supported by OTel, as it is indeed strongly typed. In general, the case was something along "if the value is not a numerical or boolean, simply convert it to string", specially useful for Exception/error objects, for example.

Author: 'Sadly' I realized even the dynamic OTel implementations do check for the type upon setting the `Attribute`, so I'd rather do a conversion at this moment. Filled #1509 as a follow up, in case we update this part (otherwise, we would be dropping values when using the OT Shim, which is *not* something I'd expect).

to a string.

### Log

Parameters:

- A set of key/value pairs, where keys must be strings, and the values may have
  any type.

Calls `Add Event` on the underlying OpenTelemetry `Span` with the specified
key/value pair set.

The `Add Event`'s `name` parameter MUST be the value with the `event` key in

> the pair set, or else fallback to use the `log` literal string.

Reviewer: Where are we with OTEL error/exception reporting? I.e. do we have proper APIs or do we only rely on semantic conventions? 

OT has a specific semantic convention about how errors are supposed to be reported (e.g. `event=error`, etc.), that seems like a conversion should be taking place here to map to OTEL, either via semconv or via proper Errors API.

Author: We don't rely on semantic conventions at this moment - we rely on a `Status` object at this time.

So, yes - my impression is that we can the `error` events 'go-through' our `AddEvent` operation. Do you have something else in mind? My impression is that the **actual** way to properly set an error is through `span.setTag("error", true)`, which we handle properly at this moment.

Author: I suggest we do this in a follow-up, as in theory we should be able to 'translate' somehow a specific `Log(Exception)` call to a `RecordException(Exception)` call, but the details need to be analyzed.

The set of values MUST be [mapped](#semantic-conventions-mapping) to the respective

> OpenTelemetry `Attribute`, or converted to a string if the value type is not supported.

Reviewer: that section does not contain anything for logs, I would not reference it. If we do need the mapping for logs lets include it here.

If an explicit timestamp is specified, a conversion MUST be done to match the
OpenTracing and OpenTelemetry units.

### Finish

Calls `End` on the underlying OpenTelemetry `Span`.

If an explicit timestamp is specified, a conversion MUST be done to match the
OpenTracing and OpenTelemetry units.

> ### SpanContext Shim

Reviewer: move to a section at the same level as Span Shim

The `SpanContext` Shim MUST contain the associated `Span` and `Baggage` values.

This object MUST be immutable.

#### Get Baggage Items

> Returns a Shim object exposing the associated OpenTelemetry `Baggage` values.

Reviewer: "Shim" is a bit weird here, in all OT APIs this either returns an iterator or a well-defined collection.

```suggestion
Returns a dictionary, collection, or iterator (depending on the requirements of the OpenTracing API for a specific language)
backed by the associated OpenTelemetry `Baggage` values.
```

Author: In OT Java we do need a Shim, but that may be a specific case. Adjusting now, thanks.

## ScopeManager Shim

For OpenTracing languages implementing the `ScopeManager` interface,  its operations
MUST be implemented using the OpenTelemetry Context Propagation API in order
to get and set active `Context` instances.

### Activate a Span

Parameters:

- A `Span`.

Gets the [associated](#opentelemetry-span-and-spancontext-relationship)
`SpanContext` Shim for the specified `Span` and puts its OpenTelemetry

> `SpanContext` and `Bagagge` in a new `Context`, which is then set as the

Reviewer: iirc the OTEL Context stores OTEL Span, not Span Context.

Author: Yes, definitely, thanks ;) I totally oversaw this.

currently active instance.

### Get the active Span

> Gets the active OpenTelemetry `Span` and returns a `Span` Shim wrapping it.

Reviewer: This is a tricky one. Are we saying a new Span Shim must be constructed at this point, or that the Span Shim should be stored in the Context? I think for performance reasons we want the latter.

Author: Try to put a `SpanShim` into a `Context` is an interesting optimization. Still, for cases where activation didn't happen _from_ the OT side, we would still need to create it:

```java
static void myMethod1() {
  // OTel api
  try (Scope scope = span.makeCurrent()) {
    myMethod2();
  }
}

static void myMethod2() {
  // No way OTel can know about the shim.
  io.opentracing.Span span = GlobalTracer.get().activeSpan();
}
```

As it sounds like an easy optimization (trying to put/get the `SpanShim` in the `Context` by default), I can definitely add it, if that sounds ok. Please confirm if this sounds fine with you.

# OpenTracing Compatibility

<details>
<summary>Table of Contents</summary>

> 

Reviewer 2: Please add status

This is accomplished by getting the
[associated](#opentelemetry-span-and-spancontext-relationship)
`SpanContext` Shim and do a lookup for the specified key in the OpenTelemetry
`Baggage` instance.

### Set Baggage Item

Parameters:

- The baggage key, a string.
- The baggage value, a string.

Creates a new `SpanContext` Shim with a new OpenTelemetry `Baggage` containing
the specified `Baggage` key/value pair. The resulting `SpanContext` Shim is then
[associated](#opentelemetry-span-and-spancontext-relationship) to the underlying
OpenTelemetry `Span`.

### Set Tag

Parameters:

- The tag key, a string.
- The tag value, which must be either a string, a boolean value, or a numeric type.

Calls `Set Attribute` on the underlying OpenTelemetry `Span` with the specified
key/value pair.

The value MUST be [mapped](#semantic-conventions-mapping) to the respective
OpenTelemetry `Attribute`, or converted to a string if the value type is not supported.

If the type of the specified value is not supported, the value MUST be converted
to a string.

### Log

Parameters:

- A set of key/value pairs, where keys must be strings, and the values may have
  any type.

> Calls `Add Event` on the underlying OpenTelemetry `Span` with the specified

Reviewer 3: minor: consider `Add Events` (plural form) to align with the [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#add-events)?

key/value pair set.

The `Add Event`'s `name` parameter MUST be the value with the `event` key in
the pair set, or else fallback to use the `log` literal string.

> The set of values MUST be [mapped](#semantic-conventions-mapping) to the respective

Reviewer 3: Here we didn't specify what to do if the value passed in via the Shim is null (assuming OpenTracing will discard the key/value pair if value is null?).

Author: Oh, good call. I think we should filter them out, to stay defensive. Will fill a follow up for that so the entire doc is reviewed for that part.

OpenTelemetry `Attribute`, or converted to a string if the value type is not supported.

If an explicit timestamp is specified, a conversion MUST be done to match the
OpenTracing and OpenTelemetry units.

### Finish

Calls `End` on the underlying OpenTelemetry `Span`.

If an explicit timestamp is specified, a conversion MUST be done to match the
OpenTracing and OpenTelemetry units.

### SpanContext Shim

The `SpanContext` Shim MUST contain the associated `Span` and `Baggage` values.

This object MUST be immutable.

#### Get Baggage Items

Returns a dictionary, collection, or iterator (depending on the requirements of the OpenTracing API for a specific language)
backed by the associated OpenTelemetry `Baggage` values.

## ScopeManager Shim

> For OpenTracing languages implementing the `ScopeManager` interface,  its operations

Reviewer 3: nit:

```suggestion
For OpenTracing languages implementing the `ScopeManager` interface, its operations
```

* [Abstract](#abstract)
* [Create an OpenTracing Tracer Shim](#create-an-opentracing-tracer-shim)
* [Tracer Shim](#tracer-shim)
  * [Inject](#inject)
  * [Extract](#extract)
* [Span Shim](#span-shim)
  * [Get SpanContext Shim](#get-spancontext-shim-for-opentelemetry-span)
  * [Set SpanContext Shim](#et-spancontext-shim-for-opentelemetry-span)
  * [Get Context](#get-context)
  * [Get Baggage Item](#get-baggage-item)
  * [Set Baggage Item](#set-baggage-item)
  * [Set Tag](#set-tag)
  * [Log](#log)
  * [Finish](#finish)
  * [SpanContext Shim](#spancontext-shim)
    * [Get Baggage Items](#get-baggage-items)
* [ScopeManager Shim](#scopemanager-shim)
  * [Activate a Span](#activate-a-span)
  * [Get the active Span](#get-the-active-span)

</details>

## Abstract

The OpenTelemetry project aims to provide backwards compatibility with the
[OpenTracing](https://opentracing.io) project in order to ease migration of
instrumented codebases.

This functionality will be provided as a bridge layer implementing the
[OpenTracing API](https://github.com/opentracing/specification) using the
OpenTelemetry API. This layer MUST NOT rely on implementation specific details

> of any SDK.

Reviewer 4: should shim allow both - OT and Otel to work simultaneously for a transition period? I don't think so, but softly. Perhaps it can be said explicitly

Author: I'm _slightly_ uncomfortable with supporting such scenario, but this is something users are actually asking for, it seems. Yes, worth mentioning it here.

Author: Added this part ;)

More specifically, the intention is to allow OpenTracing instrumentation to be
recorded using the OpenTelemetry API. This Shim Layer MUST NOT allow older

> OpenTracing Tracing components to consume upstream OpenTelemetry functionality.

Reviewer 5: I couldn't follow this sentence about upstream functionality.

Author: Amended, let me know - essentially we don't want the OpenTracing Shim to expose in its pubic API any upstream OpenTelemetry functionality, i.e. `Sampler` or `Context`.

> This functionality MUST defined in its own OpenTracing Shim Layer, not in the

Reviewer 4: ```suggestion
This functionality MUST be defined in its own OpenTracing Shim Layer, not in the
```

  context injection and extraction. Usually these are the global
  `Composite Propagator`s.

The API MUST return an OpenTracing `Tracer`.

## Tracer Shim

### Inject

Parameters:

- A `SpanContext`.
- A `Format` descriptor.
- A carrier.

Inject the underlying OpenTelemetry `Span` and `Bagagge` using the registered
OpenTelemetry `Propagator`s:

- `TEXT_MAP` and `HTTP_HEADERS` formats MUST use the registered OpenTelemetry
  `HTTPTextPropagator`, if any.

Errors MUST NOT be raised if the specified `Format` is `BINARY`.

### Extract

Parameters:

- A `Format` descriptor.
- A carrier.

Extract OpenTelemetry `Span` and `Baggage` from a carrier using the registered
OpenTelemetry `Propagator`s:

- `TEXT_MAP` and `HTTP_HEADERS` formats MUST use the registered OpenTelemetry
  `HTTPTextPropagator`, if any.

Returns a `SpanContext` Shim with the underlying extracted OpenTelemetry
`Span` and `Baggage`, or null if either the `Format` is `BINARY` or
no value could be extracted.

## Span Shim

The `Span` operations MUST be implemented using underlying OpenTelemetry `Span`
and `Baggage` values through a `SpanContext` Shim.

In order to satisfy the OpenTracing `Span` requirements:

> - The associated `SpanContext` Shim object will contain a `Context`, which

Reviewer: > will contain a `Context`

do you mean OpenTelemetry `Context`?

Reviewer: Also, why is it a requirement? The shim can extract OTEL Span and Baggage if it needs to.

Author: So originally added this way as it made it simpler to do activation in a single step, as we already have a `Context`, but I'm having second thoughts about this, as we may leak other values in `Context`. 

Overall I agree that `Span` and `Baggage` can be extracted instead. Sounds like a plan?

Author: Updated to have `Span` and `Baggage` extracted instead.

  will contain the OpenTelemetry `Span` and `Baggage` values.
  This `SpanContext` Shim MUST be immutable and MUST be replaced every time

>   baggage is updated through `SetBaggageItem`.

Reviewer: SetBaggageItem is never mentioned again in this doc, while it is probably the hardest part of the compatibility because of the inverted relationship between tracer and context propagation in OT vs OTEL.

Author: Fixed the link, my bad ;(

- An underlying OpenTelemetry `Span` MUST be associated with only one
  `SpanContext` Shim at a time for all execution units. This is done in order
  to keep its baggage values consistent across all execution units at all
  times.

This is a simple graphical representation of the mentioned objects:

```
  Span Shim
  +- OpenTelemetry Span
     +- SpanContext Shim
        +- OpenTelemetry Context
           +- OpenTelemetry Span
           +- OpenTelemetry Baggage
```

The OpenTelemetry `Span` is contained in `Span` Shim in order to easily
get and set its currently associated `SpanContext` Shim through
[Get Span Context Shim](#get-spancontext-shim-for-opentelemetry-span) and
[Set Span Context Shim](#set-spancontext-shim-for-opentelemetry-span).

The OpenTelemetry `Span` is also contained in the related OpenTelemetry
`Context` in order to simplify setting the active OpenTelemetry `Span` and
`Baggage` in a single step.

> The `Log` operations MUST be implemented using the OpenTelemetry

Reviewer 4: is there some mapping between well known OT logs and OpenTelemetry semantic convention?

Author: We only have `error`, which happens at `SetTag`, so I will add it. There were plans to add other members, but that can be done in a follow up (or after GA even).

Author: Added it as a section at the end (currently we only have `error` - we can add more mappings later on).

`Span`'s `Add Events` operations.

> The `Set Tag` operations MUST be implemented using the OpenTelemetry

Reviewer 4: same qq as for logs - do we need to document any mapping between well-known tags and attributes?

Author: Same as above.

`Span`'s `Set Attributes` operations.

> ### Get SpanContext Shim for OpenTelemetry Span

Reviewer: This section does not follow from the document structure. Is it referring to operations of the span shim? Is this some kind of public API of the shim?

Author: This one and `Set SpanContext Shim for OpenTelemetry Span` are supposed to be Internal operations, i.e. non-public operations:

> This is an INTERNAL helper operation, used by the Shim Layer only. 

... and are defined in order to abstract this step - in turn, this is because each language will decide how to internally track the link between a given OTel Span and a OT SpanContext:

- In Java, we can leverage a `Map` (as shown in the current OT Shim).
- In Python, the OTel Span can have its linked OT SpanContext as `Span._ot_span_context`.

If you think we should make this operations more clear, let me know (maybe add an `Internal` prefix?)

Reviewer: If it's internal why does it need to be spelled out?  

> how to internally track the link between a given OTel Span and a OT SpanContext

Why is this necessary to track?

Reviewer: also, if this truly necessary, I would rather have some ability to attach arbitrary runtime-only payloads to OTEL Span than to direct people to use a global map / registry, which are difficult to manage.

Author: > I would rather have some ability to attach arbitrary runtime-only payloads to OTEL Span than to direct people to use a global map / registry, which are difficult to manage.

For the record, I had initially [proposed this](https://github.com/open-telemetry/opentelemetry-java/pull/596) but this was deemed a bad approach (I agree that handling a global registry can become tricky). We can revisit this as needed for Java and other languages.

> Why is this necessary to track?

Fair enough. Will 'inline' them in the related places.

Author: Removed them with a very general note on the OTel Span / OT SpanContext relationship.

## Tracer Shim

### Inject

Parameters:

- A `SpanContext`.
- A `Format` descriptor.
- A carrier.

Inject the underlying OpenTelemetry `Span` and `Bagagge` using the registered
OpenTelemetry `Propagator`s:

- `TEXT_MAP` and `HTTP_HEADERS` formats MUST use the registered OpenTelemetry
  `HTTPTextPropagator`, if any.

Errors MUST NOT be raised if the specified `Format` is `BINARY`.

### Extract

Parameters:

- A `Format` descriptor.
- A carrier.

Extract OpenTelemetry `Span` and `Baggage` from a carrier using the registered
OpenTelemetry `Propagator`s:

- `TEXT_MAP` and `HTTP_HEADERS` formats MUST use the registered OpenTelemetry
  `HTTPTextPropagator`, if any.

Returns a `SpanContext` Shim with the underlying extracted OpenTelemetry
`Span` and `Baggage`, or null if either the `Format` is `BINARY` or
no value could be extracted.

## Span Shim

The `Span` operations MUST be implemented using underlying OpenTelemetry `Span`
and `Baggage` values through a `SpanContext` Shim.

In order to satisfy the OpenTracing `Span` requirements:

- The associated `SpanContext` Shim object will contain a `Context`, which
  will contain the OpenTelemetry `Span` and `Baggage` values.

>   This `SpanContext` Shim MUST be immutable and MUST be replaced every time

Reviewer: doesn't seem to be the case here: https://github.com/open-telemetry/opentelemetry-python/blob/cd63eb6e52c6f66c9170a2928d2b4dc8918a41de/instrumentation/opentelemetry-instrumentation-opentracing-shim/src/opentelemetry/instrumentation/opentracing_shim/__init__.py#L294

Author: Correct, this needs to be fixed (similarly how to we need to support the `timestamp` values in the Java one). But yes, we need to definitely make the OT `SpanContext` part immutable.
